src/api/baggage.rs

```diff
@@ -112,35 +162,53 @@ impl FromIterator<(Key, Value)> for Baggage {
 impl FromIterator<KeyValue> for Baggage {
     fn from_iter<I: IntoIterator<Item = KeyValue>>(iter: I) -> Self {
         Baggage {
-            inner: iter.into_iter().map(|kv| (kv.key, kv.value)).collect(),
+            inner: iter
+                .into_iter()
+                .map(|kv| (kv.key, (kv.value, Metadata::default())))
+                .collect(),
+        }
+    }
+}
+
+impl FromIterator<KeyValueMetadata> for Baggage {
+    fn from_iter<I: IntoIterator<Item = KeyValueMetadata>>(iter: I) -> Self {
+        Baggage {
+            inner: iter
+                .into_iter()
+                .map(|kvm| (kvm.key, (kvm.value, kvm.metadata)))
+                .collect(),
         }
     }
 }
 
 /// Methods for sorting and retrieving baggage data in a context.
 pub trait BaggageExt {
-    /// Returns a clone of the current context with the included name / value pairs.
+    /// Returns a clone of the given context with the included name / value pairs.
     ///
     /// # Examples
     ///
     /// ```
     /// use opentelemetry::api::{BaggageExt, Context, KeyValue, Value};
     ///
-    /// let cx = Context::current_with_baggage(vec![KeyValue::new("my-name", "my-value")]);
+    /// let cx = Context::current().with_cleared_baggage();
     ///
-    /// assert_eq!(
-    ///     cx.baggage().get("my-name"),
-    ///     Some(&Value::String("my-value".to_string())),
-    /// )
+    /// assert_eq!(cx.baggage().len(), 0);
     /// ```
-    fn current_with_baggage<T: IntoIterator<Item = KeyValue>>(baggage: T) -> Self;
+    fn with_cleared_baggage(&self) -> Self;
 
+    /// Returns a reference to this context's baggage, or the default
+    /// empty baggage if none has been set.
+    fn baggage(&self) -> &Baggage;
+}
+
+/// Method to retrieve key value pairs, possible along with the metadata.
+pub trait AddBaggage<I> {
```

Comment:
I think you could get away without the extra `AddBaggage` trait and merge everything into the `BaggageExt` trait using `I: Into<KeyValueMetadata>`. That works for both iterators over `KeyValue` and `KeyValueMetadata`. Longer example:

```rust
pub trait BaggageExt {
    fn with_cleared_baggage(&self) -> Self;
    fn baggage(&self) -> &Baggage;
    fn with_baggage<T: IntoIterator<Item = I>, I: Into<KeyValueMetadata>>(&self, baggage: T) -> Self;
    fn current_with_baggage<T: IntoIterator<Item = I>, I: Into<KeyValueMetadata>>(baggage: T) -> Self;
}

impl BaggageExt for Context {
    // ...

    fn with_baggage<T: IntoIterator<Item = I>, I: Into<KeyValueMetadata>>(&self, baggage: T) -> Self {
        let merged: Baggage = self
            .baggage()
            .iter()
            .map(|(key, (value, metadata))| {
                KeyValueMetadata::new(key.clone(), value.clone(), metadata.clone())
            })
            .chain(baggage.into_iter().map(|kvs| kvs.into()))
            .collect();

        self.with_value(merged)
    }

    fn current_with_baggage<T: IntoIterator<Item = I>, I: Into<KeyValueMetadata>>(baggage: T) -> Self {
        Context::current().with_baggage(baggage)
    }
}
```

Reply:
Good advice! It looks more elegant.